ports/rp2/modrp2.c

```diff
@@ -50,6 +50,59 @@ STATIC mp_obj_t rp2_country(size_t n_args, const mp_obj_t *args) {
 STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(rp2_country_obj, 0, 1, rp2_country);
 #endif
 
+#include "pico/stdlib.h"
+#include "hardware/gpio.h"
+#include "hardware/sync.h"
+#include "hardware/structs/ioqspi.h"
+#include "hardware/structs/sio.h"
+
+// Picoboard has a button attached to the flash CS pin, which the bootrom
+// checks, and jumps straight to the USB bootcode if the button is pressed
+// (pulling flash CS low). We can check this pin in by jumping to some code in
+// SRAM (so that the XIP interface is not required), floating the flash CS
+// pin, and observing whether it is pulled low.
+//
+// This doesn't work if others are trying to access flash at the same time,
+// e.g. XIP streamer, or the other core.
+// https://github.com/raspberrypi/pico-examples/blob/master/picoboard/button/button.c
+
+bool __no_inline_not_in_flash_func(get_bootsel_button)() {
+    const uint CS_PIN_INDEX = 1;
+
+    // Must disable interrupts, as interrupt handlers may be in flash, and we
+    // are about to temporarily disable flash access!
+    uint32_t flags = save_and_disable_interrupts();
+
+    // Set chip select to Hi-Z
+    hw_write_masked(&ioqspi_hw->io[CS_PIN_INDEX].ctrl,
+        GPIO_OVERRIDE_LOW << IO_QSPI_GPIO_QSPI_SS_CTRL_OEOVER_LSB,
+            IO_QSPI_GPIO_QSPI_SS_CTRL_OEOVER_BITS);
+
+    // Note we can't call into any sleep functions in flash right now
+    for (volatile int i = 0; i < 1000; ++i) {;
```

Comment:
This loop with a count of 1000 seems rather arbitrary.  Does it work in all cases?  Does it depend on CPU frequency? Can it be made shorter?

Reply:
FWIW, this is the equivalent code in the boot rom for detecting the bootsel button:
https://github.com/raspberrypi/pico-bootrom/blob/master/bootrom/bootrom_main.c#L243

Perhaps we should adapt that (both the delay times, and the 5/9 count).

Reply:
> // This doesn't work if others are trying to access flash at the same time,
> // e.g. XIP streamer, or the other core.

Can we use the `multicore_lockout_start_blocking` functionality to make this work? We already use that elsewhere when we need to prevent the other core from accessing flash.

i.e. if you wrapped this whole function with

```
mp_uint_t atomic_state = MICROPY_BEGIN_ATOMIC_SECTION();
...
MICROPY_END_ATOMIC_SECTION(atomic_state);
```

then you'd get the lockout automatically.